internal/proxy/simple_rate_limiter.go

```diff
@@ -0,0 +1,344 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package proxy
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"sync"
+
+	"go.uber.org/zap"
+
+	"github.com/milvus-io/milvus-proto/go-api/v2/commonpb"
+	"github.com/milvus-io/milvus-proto/go-api/v2/milvuspb"
+	"github.com/milvus-io/milvus/internal/proto/internalpb"
+	"github.com/milvus-io/milvus/internal/proto/proxypb"
+	"github.com/milvus-io/milvus/internal/util/quota"
+	rlinternal "github.com/milvus-io/milvus/internal/util/ratelimitutil"
+	"github.com/milvus-io/milvus/pkg/config"
+	"github.com/milvus-io/milvus/pkg/log"
+	"github.com/milvus-io/milvus/pkg/metrics"
+	"github.com/milvus-io/milvus/pkg/util/paramtable"
+	"github.com/milvus-io/milvus/pkg/util/ratelimitutil"
+	"github.com/milvus-io/milvus/pkg/util/typeutil"
+)
+
+// SimpleLimiter is implemented based on Limiter interface
+type SimpleLimiter struct {
+	quotaStatesMu sync.RWMutex
+	rateLimiter   *rlinternal.RateLimiterTree
+}
+
+// NewSimpleLimiter returns a new SimpleLimiter.
+func NewSimpleLimiter() *SimpleLimiter {
+	rootRateLimiter := newClusterLimiter()
+	m := &SimpleLimiter{rateLimiter: rlinternal.NewRateLimiterTree(rootRateLimiter)}
+	return m
+}
+
+// Check checks if request would be limited or denied.
+func (m *SimpleLimiter) Check(dbID int64, collectionIDToPartIDs map[int64][]int64, rt internalpb.RateType, n int) error {
+	if !Params.QuotaConfig.QuotaAndLimitsEnabled.GetAsBool() {
+		return nil
+	}
+
+	m.quotaStatesMu.RLock()
+	defer m.quotaStatesMu.RUnlock()
+
+	// 1. check global(cluster) level rate limits
+	clusterRateLimiters := m.rateLimiter.GetRootLimiters()
+	ret := clusterRateLimiters.Check(rt, n)
+
+	if ret != nil {
+		clusterRateLimiters.Cancel(rt, n)
+		return ret
+	}
+
+	// store done limiters to cancel them when error occurs.
+	doneLimiters := make([]*rlinternal.RateLimiterNode, 0)
+	doneLimiters = append(doneLimiters, clusterRateLimiters)
+
+	cancelAllLimiters := func() {
+		for _, limiter := range doneLimiters {
+			limiter.Cancel(rt, n)
+		}
+	}
+
+	// 2. check database level rate limits
+	if ret == nil {
+		dbRateLimiters := m.rateLimiter.GetOrCreateDatabaseLimiters(dbID, newDatabaseLimiter)
+		ret = dbRateLimiters.Check(rt, n)
+		if ret != nil {
+			cancelAllLimiters()
+			return ret
+		}
+		doneLimiters = append(doneLimiters, dbRateLimiters)
+	}
+
+	// 3. check collection level rate limits
+	if ret == nil && len(collectionIDToPartIDs) > 0 && !isNotCollectionLevelLimitRequest(rt) {
+		for collectionID := range collectionIDToPartIDs {
+			// only dml and dql have collection level rate limits
+			collectionRateLimiters := m.rateLimiter.GetOrCreateCollectionLimiters(dbID, collectionID,
+				newDatabaseLimiter, newCollectionLimiters)
+			ret = collectionRateLimiters.Check(rt, n)
+			if ret != nil {
+				cancelAllLimiters()
+				return ret
+			}
+			doneLimiters = append(doneLimiters, collectionRateLimiters)
+		}
+	}
+
+	// 4. check partition level rate limits
```

Comment:
is there a specific reason we need partiton level metrics?
Not sure how much time it cost by simply check all the limiters

Reply:
The partition here is the partition involved in the request, and the corresponding limiter is obtained directly through the map, so the check should not consume a lot of resources.